libraries/CryptoInterface/src/CryptoInterface.cpp

```diff
@@ -0,0 +1,567 @@
+/*
+ * BearSSL Copyright (c) 2016 Thomas Pornin <pornin@bolet.org>
+ * Rest of this file Copyright (C) 2019 Anders Löfgren
+ *
+ * License (MIT license):
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "CryptoInterface.h"
+#include <TypeConversion.h>
+
+#include <bearssl/bearssl.h>
+#include <assert.h>
+
+namespace TypeCast = esp8266::TypeConversion;
+
+namespace experimental
+{
+  namespace
+  {
+    size_t _ctMinDataLength = 0;
+    size_t _ctMaxDataLength = 1024;
+    
+    br_hkdf_context _storedHkdfContext;
+    bool _hkdfContextStored = false;
+
+    uint8_t *defaultNonceGenerator(uint8_t *nonceArray, const size_t nonceLength)
```

Comment:
Great attention to detail on the lack of entropy.  I think a proper "get random number" call (maybe renaming it or something) would also be handy, returning just safe u32?  I can see it being handy everywhere `rand()` or `RANDOM_REG` is (improperly) used.  This fcn would also use it to grab the RANDOM_REG, of course, so that the timing would be preserved...

Reply:
If I understand this correct, you would like a version of the `defaultNonceGenerator` function that returns an uint32_t instead of an uint8_t array? I'm thinking it may be possible to break out the `defaultNonceGenerator` functionality into a separate base function that looks exactly the same as  `defaultNonceGenerator` does now. This base function can then be called from two wrapper functions, one being `defaultNonceGenerator` and the other being the new function you describe that returns an uint32_t. Was this what you had in mind?

Reply:
I'd go further.  The standard random generator(Arduino `random`) at least.

Also, I'd suggest because this is useful *everywhere* that "nonce" is probably not the right naming.  `uint32_t ESP.random()` might be the right spot to expose it.  Want to make it obvious that this is the right way to get real random #s (as opposed to newlib pseudorand which cannot be changed).

Reply:
New ESP.random methods have been added to the core based on the crypto code, and these new random methods are now used by the defaultNonceGenerator.